20240627-BC-listapi.md

```diff
@@ -0,0 +1,182 @@
+
+List API requests on state stores that don’t support the List API will result in errors.
+
+## Impact of the List API on Dapr state store components
+From the moment this proposal is accepted, all state store components will be required to implement the List API in order to get the "Stable" certification level.
+Components that are currently stable and for which the underlying state store does not support listing will not lose their stable status.
+
+## Performance and pricing implications
+Listing keys in big data sets, specially for partitioned databases, can be expensive in terms of both performance and cost. Often it would incur creating an index which will impact write performance, storage cost and sometimes even read performance.
+For the databases where this is a concern, we should offer an option to disable the List API on the component level.
+
+## Definitions:
+
+- **Listing**: The ability to retrieve a collection of items.
+- **Sorting**: The ability to sort the results based on one or more fields.
+- **Prefix Search**: The ability to search for items that start with a given prefix.
+- **Pagination**: The ability to paginate through items, typically using skip/limit or similar mechanisms.
+
+## Adendum
+
+Here's a list of the relevant capabilities of all the stable state stores:
+
+| Store | Cursor listing | Offset listing | Sorting | Number of Items per Page | Prefix Search | Comments |
+| --- | --- | --- | --- | --- | --- | --- |
+| **aws dynamodb** | Yes | No | Yes, with a GSI | Yes | Yes, with an additional sortKey and a GSI | In order to be able to use prefix search, users will need to have a Global Search Index(GSI) where the partition key will be a single fixed string (for ex. the `_` character) and the sort key will be the key name. There are some drawbacks to this that can be discussed in detail elsewhere. |
+| **azure blob store** | Yes (continuation token) | No | Always sorted in ASC order. Desc, or unsorted is not possible. | Yes | Yes | Results are always sorted by key name in ascending order. |
+| **azure cosmos db** | Yes | Yes | Yes | Yes | Yes |   |
+| **azure table storage** | Yes | No | Yes, just ASC | Yes, with $top | Yes, with range search | Partition key is the application id. |
+| **cassandra** | Yes | No | No | Yes | No | Can’t prefix search and sort across all partitions. We could consider maintaining a new table containing all keys, and mirroring the original key’s ttl. |
+| **cockroachdb** | Yes, if sorting is required | Yes | Yes | Yes | Yes | Need to create an index on the search column |
+| **gcp firestore** | Yes |   |   |   |   |   |
+| **in-memory** | No | No | No | No | No | We can implement all the features, but it’s not trivial to aggregate data across multiple instances |
```

Comment:
We don't need to make in-memory to be distributed. It has always been a local store for testing. Yes, we can implement anything we want here, but all local IMO.